streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java

```diff
@@ -102,4 +140,230 @@ static class RawAndDeserializedValue<ValueType> {
+            ValueAndTimestamp<V> value = deserializer.apply(rawResult.getResult());
+            V res = null;
+            if (value != null) {
+                res = value.value();
+            }
+            final QueryResult<V> typedQueryResult =
+                    InternalQueryResultUtil.copyAndSubstituteDeserializedResult(rawResult, res);
+            result = (QueryResult<R>) typedQueryResult;
+        } else {
+            // the generic type doesn't matter, since failed queries have no result set.
+            result = (QueryResult<R>) rawResult;
+        }
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <R> QueryResult<R> runRangeQuery2(final Query<R> query,
+                                               final PositionBound positionBound,
+                                               final QueryConfig config) {
+
+        final QueryResult<R> result;
+        final RangeQuery<K, V> typedQuery = (RangeQuery<K, V>) query;
+        final RangeQuery<Bytes, byte[]> rawRangeQuery;
+        if (typedQuery.getLowerBound().isPresent() && typedQuery.getUpperBound().isPresent()) {
+            rawRangeQuery = RangeQuery.withRange(
+                    keyBytes(typedQuery.getLowerBound().get()),
+                    keyBytes(typedQuery.getUpperBound().get())
+            );
+        } else if (typedQuery.getLowerBound().isPresent()) {
+            rawRangeQuery = RangeQuery.withLowerBound(keyBytes(typedQuery.getLowerBound().get()));
+        } else if (typedQuery.getUpperBound().isPresent()) {
+            rawRangeQuery = RangeQuery.withUpperBound(keyBytes(typedQuery.getUpperBound().get()));
+        } else {
+            rawRangeQuery = RangeQuery.withNoBounds();
+        }
+        final QueryResult<KeyValueIterator<Bytes, byte[]>> rawResult =
+                wrapped().query(rawRangeQuery, positionBound, config);
+        if (rawResult.isSuccess()) {
+            final KeyValueIterator<Bytes, byte[]> iterator = rawResult.getResult();
+            final KeyValueIterator<K,V> resultIterator = new MeteredKeyValueTimestampedIterator(
+                    iterator,
+                    getSensor,
+                    getDeserializeValue(serdes, wrapped()),
+                    true
+            );
+            final QueryResult<KeyValueIterator<K, V>> typedQueryResult =
+                    InternalQueryResultUtil.copyAndSubstituteDeserializedResult(
+                            rawResult,
+                            resultIterator
+                    );
+            result = (QueryResult<R>) typedQueryResult;
+        } else {
+            // the generic type doesn't matter, since failed queries have no result set.
+            result = (QueryResult<R>) rawResult;
+        }
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    private class MeteredKeyValueTimestampedIterator implements KeyValueIterator<K, V> {
```

Comment:
Seems this is copy of existing `MeteredKeyValueTimestampedIterator` from `MeteredKeyValueStore`.

I think we should not just make a duplication of the code, but implement is in a way to re-use code (in the end, the difference is only that we need an additional "unwrapping", that I think we can get done eg by making it part of the `Function` that we pass in.

Side note: as discussed in person, the original class should be renamed -- not sure why it has `Timestamped` in it's name -- does not make any sense.

Reply:
ok

Reply:
Seems this comment was addressed yet?

Reply:
Yes, we reuse the code, but the name not change yet.